src/PGP/datasets/visual_data_helper.py

```diff
@@ -0,0 +1,183 @@
+        if final_coords is None:
+            return None
+        else:
+            min_x, min_y, max_x, max_y = final_coords
+            return {
+                'cam_channel': sd_rec['channel'],
+                'cam_file': sd_rec['filename'],
+                'roi_min_x': min_x,
+                'roi_min_y': min_y,
+                'roi_max_x': max_x,
+                'roi_max_y': max_y,
+            }
+        
+    def _best_channel(self, ann_rec, channel_pref=None):
+        """
+        Returns the best camera to view ann_tok from
+            If no camera can view ann_tok (yes, this can happen, idk why), returns None
+
+        ann_tok: Annotation token to get the best view of
+        channel_pref: If you can view ann_tok from this channel, do so
+        """
+        # ann_rec = self.nusc.get('sample_annotation', ann_tok)
+        sample_rec = self.nusc.get('sample', ann_rec['sample_token'])
+        channels = [key for key in sample_rec['data'].keys() if 'CAM' in key]
+        views = []
+        for channel in channels:
+            view = self._camera_box(sample_rec['data'][channel], ann_rec['token'])
+            if view:
+                views.append(view)
+        if len(views) == 0:
+            return None
+        if channel_pref is not None:
+            pref_view = [v for v in views if v['cam_channel'] == channel_pref]
+            if len(pref_view) > 0:
+                assert len(pref_view) == 1
+                return pref_view[0]
+        def area(view):
+            return (view['roi_max_x'] - view['roi_min_x']) * (view['roi_max_y'] - view['roi_min_y'])
+        return sorted(views, key=area, reverse=True)[0]
+    
+    def get_target_agent_visuals(self, s_t: str, i_t: str, seconds: int) -> List[Dict]:
+        curr_ann_tok = self._ann_token_from_sample_instance(s_t, i_t)
+        ann_rec = self.nusc.get('sample_annotation', curr_ann_tok)
+        ann_rec_hist = [ann_rec]
+        # nuscenes dataset is annotated at 2Hz
+        for _ in range(2 * seconds):
+            # if we've ran out of history for this instance
+            if not ann_rec['prev']:
+                break
+            ann_rec = self.nusc.get('sample_annotation', ann_rec['prev'])
+            ann_rec_hist.append(ann_rec)
+        curr_channel = None
+        views = []
+        for ar in ann_rec_hist:
+            view = self._best_channel(ar, curr_channel)
+            views.append(view)
+            if view:
+                curr_channel = view['cam_channel']
+        return views
+    
+    def load_visuals(self, data: Dict):
+        if self.transform is None:
+            return data
+        visuals = data['inputs']['target_agent_visuals']
+        imgs = [Image.open(f"{self.nusc.dataroot}/{v['cam_file']}") for v in visuals]
+        imgs = [img.convert("RGB") for img in imgs]
+        data['inputs']['target_agent_images'] = \
+            torch.stack([self.transform(img) for img in imgs])
```

Comment:
Since we have a variable number of images (might not have the full 2 second history of the agent), this returns a non-constant sized tensor (in its first dim)...and thus the resulting datapoints cannot be batched (the error that we ran into last night).

How the PGP author got around this (since he ran into the same thing with his non-constant size geometry feature sequences) is to pad them with zeros, and include a `valid` mask. I'm going to look into if we can apply the same technique to get around our issue.